src/Juvix/Compiler/Backend/Geb/Evaluator.hs

```diff
@@ -0,0 +1,377 @@
+        OpSub -> return $ GebValueMorphismInteger $ l - r
+        OpMul -> return $ GebValueMorphismInteger $ l * r
+        OpDiv -> return $ GebValueMorphismInteger $ l `div` r
+        OpMod -> return $ GebValueMorphismInteger $ l `mod` r
+        OpLt ->
+          if
+              | l < r -> return valueTrue
+              | otherwise -> return valueFalse
+        OpEq ->
+          if
+              | l < r -> return valueTrue
+              | otherwise -> return valueFalse
+    (m1, m2) -> case binop ^. binopOpcode of
+      OpEq ->
+        if
+            | sameKind m1 m2 ->
+                if
+                    | m1 == m2 -> return valueTrue
+                    | otherwise -> return valueFalse
+            | otherwise ->
+                throw
+                  EvalError
+                    { _evalErrorMsg = "Equality can only be applied to values of the same kind.",
+                      _evalErrorGebValue = Just (lfPair m1 m2),
+                      _evalErrorGebExpression = (Just (MorphismBinop binop))
+                    }
+      _ ->
+        throw
+          EvalError
+            { _evalErrorMsg = "Cannot apply operation",
+              _evalErrorGebValue = Just (lfPair m1 m2),
+              _evalErrorGebExpression = Just (MorphismBinop binop)
+            }
+
+sameKind :: GebValue -> GebValue -> Bool
+sameKind l r = case (l, r) of
+  (GebValueMorphismInteger _, GebValueMorphismInteger _) -> True
+  (GebValueMorphismUnit, GebValueMorphismUnit) -> True
+  (GebValueMorphismLeft _, GebValueMorphismLeft _) -> True
+  (GebValueMorphismRight _, GebValueMorphismRight _) -> True
+  (GebValueMorphismPair _, GebValueMorphismPair _) -> True
+  (GebValueClosure _, GebValueClosure _) -> True
+  _ -> False
+
+valueTrue :: GebValue
+valueTrue = GebValueMorphismLeft GebValueMorphismUnit
+
+valueFalse :: GebValue
+valueFalse = GebValueMorphismRight GebValueMorphismUnit
+
+requiresObjectInfo :: GebValue -> Bool
+requiresObjectInfo = \case
+  GebValueMorphismUnit -> False
+  GebValueMorphismInteger {} -> False
+  GebValueClosure {} -> True
+  GebValueMorphismLeft {} -> True
+  GebValueMorphismRight {} -> True
+  GebValueMorphismPair {} -> True
+
+quote :: Maybe Object -> GebValue -> Sem r Morphism
+quote ty = \case
+  GebValueClosure cls -> quoteClosure ty cls
+  GebValueMorphismInteger i -> return $ MorphismInteger i
+  GebValueMorphismLeft m -> quoteValueMorphismLeft ty m
+  GebValueMorphismPair m -> quoteValueMorphismPair ty m
+  GebValueMorphismRight m -> quoteMorphismRight ty m
+  GebValueMorphismUnit -> return MorphismUnit
+
+quoteClosure :: Maybe Object -> ValueClosure -> Sem r Morphism
+quoteClosure ty cls = do
+  case ty of
+    Just (ObjectHom funObj) ->
+      return $
+        MorphismLambda
+          Lambda
+            { _lambdaVarType = funObj ^. homDomain,
+              _lambdaBodyType = funObj ^. homCodomain,
+              _lambdaBody =
+                cls ^. valueClosureLambdaBody
```

Comment:
This is incorrect. You need to substitute the free variables of the body for the values in the environment.